// react-auth-wrapper/es/Redirect.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _helpers = require("./helpers");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

var Redirect =
/*#__PURE__*/
function (_PureComponent) {
  _inheritsLoose(Redirect, _PureComponent);

  function Redirect() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = Redirect.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        redirectPath = _this$props.redirectPath,
        location = _this$props.location,
        history = _this$props.history;

    if (redirectPath) {
      if (history && typeof history.push === 'function') {
        history.push(redirectPath);
      } else if (location && typeof location.replace === 'function') {
        location.replace(redirectPath);
      } else if (typeof window !== 'undefined') {
        // eslint-disable-next-line no-undef
        window.location.replace(redirectPath);
      }
    }

    return null;
  };

  return Redirect;
}(_react.PureComponent);

var defaults = {
  authenticatedSelector: function authenticatedSelector(p) {
    return Boolean(p.isAuthenticated);
  },
  authenticatingSelector: function authenticatingSelector(p) {
    return Boolean(p.isAuthenticating);
  },
  allowRedirectBack: true,
  FailureComponent: Redirect,
  wrapperDisplayName: 'AuthRedirectWrapper',
  redirectQueryParamName: 'redirect'
};

function getLocation(props) {
  if (props === void 0) {
    props = {};
  }

  return props.location || typeof window !== 'undefined' && window.location || {};
}

function createRedirectWrapper(wrapper) {
  function withRedirect(args) {
    if (args === void 0) {
      args = {};
    }

    var allArgs = _objectSpread({}, defaults, args);

    var redirectPath = allArgs.redirectPath,
        FailureComponent = allArgs.FailureComponent,
        allowRedirectBack = allArgs.allowRedirectBack,
        authenticatedSelector = allArgs.authenticatedSelector,
        authenticatingSelector = allArgs.authenticatingSelector,
        _allArgs$redirectQuer = allArgs.redirectQueryParamName,
        redirectQueryParamName = _allArgs$redirectQuer === void 0 ? 'redirect' : _allArgs$redirectQuer;
    var allowRedirectBackFn = (0, _helpers.validateBoolOrFunction)(allowRedirectBack, 'allowRedirectBack');
    var getAuthenticated = (0, _helpers.validateStringOrFunction)(authenticatedSelector, 'authenticatedSelector');
    var getAuthenticating = (0, _helpers.validateStringOrFunction)(authenticatingSelector, 'authenticatingSelector');

    var EnhancedFailureComponent = function EnhancedFailureComponent(props) {
      return _react.default.createElement(FailureComponent, _extends({}, props, {
        redirectPath: (0, _helpers.parse)(getLocation(props).search, redirectQueryParamName) || (0, _helpers.getFullRedirectPath)(redirectQueryParamName, props.redirectPath || redirectPath, allowRedirectBackFn(props), getLocation(props))
      }));
    };

    return function (DecoratedComponent) {
      var AuthWrapped = wrapper(_objectSpread({}, allArgs, {
        FailureComponent: EnhancedFailureComponent
      }))(DecoratedComponent);
      return function (props) {
        return _react.default.createElement(AuthWrapped, _extends({}, props, {
          isAuthenticated: getAuthenticated(props),
          isAuthenticating: getAuthenticating(props)
        }));
      };
    };
  }

  return withRedirect;
}

var _default = createRedirectWrapper;
exports.default = _default;
